File: Depth_First_Search/1367.Linked_List_in_Binary_Tree/C++/1367.Linked_List_in_Binary_Tree.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isSubPath(ListNode * head, TreeNode * root)
    {
        //=================================//
        // Input type:                     //
        //   - ListNode * (head list-node) //
        //   - TreeNode * (root tree-node) //
        // Return type:                    //
        //   - bool (result flag)          //
        //=================================//
        if (! root) { return false; } //Issue, Boundary-case handling

        if (dfsTraversal(head, root)) { return true; } //Resursion function call (top-first root tree-node)

        return (isSubPath(head, (root->left)) || isSubPath(head, (root->right))); //Resursion function call (following root tree-node)
    }


    bool dfsTraversal(ListNode * head, TreeNode * root)
    {
        //=================================//
        // Input type:                     //
        //   - ListNode * (head list-node) //
        //   - TreeNode * (root tree-node) //
        // Return type:                    //
        //   - bool (result flag)          //
        //=================================//
        if (! head) { return true; } //Issue, Boundary-case handling

        if (! root) { return false; } //Issue, Boundary-case handling

        if ((head->val) != (root->val)) { return false; } //Issue, Boundary-case handling

        return (dfsTraversal((head->next), (root->left)) || dfsTraversal((head->next), (root->right))); //Resursion function call (next root tree-node)
    }
};

File: Depth_First_Search/1367.Linked_List_in_Binary_Tree/ptyhon/1367.Linked_List_in_Binary_Tree.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def isSubPath(self, head, root):
        """
        :type head: Optional[ListNode]
        :type root: Optional[TreeNode]
        :rtype: bool
        """
        #======================================#
        # Recursion-based DFS traversal method #
        #======================================#
        if (not root): #Issue/Boundary-case handling
            return False

        if (self.dfsTraversal(head, root)): #Recursion function call (i.e. top-first root tree-node)
            return True

        return (self.isSubPath(head, root.left) or self.isSubPath(head, root.right)) #Recursion function call (i.e. following root tree-node)
        

    def dfsTraversal(self, head, root):
        """
        :type head: Optional[ListNode]
        :type root: Optional[TreeNode]
        :rtype: bool
        """
        #========================================#
        # Recursion-based DFS traversal function #
        #========================================#
        if (not head): #Issue/Boundary-case handling
            return True

        if (not root): #Issue/Boundary-case handling
            return False

        if (head.val != root.val): #Issue/Boundary-case handling
            return False

        return (self.dfsTraversal(head.next, root.left) or self.dfsTraversal(head.next, root.right)) #Recursion function call (i.e. next root tree-node)
